src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs

```diff
@@ -525,23 +530,21 @@ private IEnumerable<LibraryDependency> GetLibraryDependenciesForCentralTransitiv
                         continue;
                     }
 
+                    if (dependencyDictionary == null)
+                    {
+                        dependencyDictionary = rootNode.Item.Data.Dependencies.ToDictionary(x => x.Name, x => x, StringComparer.OrdinalIgnoreCase);
```

Comment:
The fact that GraphNode doesn't implement hashcode made me a bit nervous, but I think it's ok in this case, as the data itself will be duplicated.

Reply:
Hmm, not sure what was your concern really. Was it about `HashSet<GraphNode<T>>` for visited nodes? I think it is ok that it doesn't have proper equality comparer. We care about not visiting the same graph node twice, but it is ok when there are multiple nodes representing same project or package.